libq/cur_sys_pkg.c

```diff
@@ -0,0 +1,348 @@
+#include <dirent.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <xalloc.h>
+
+#include "atom.h"
+#include "hash.h"
+#include "xchdir.h"
+#include "cur_sys_pkg.h"
+
+
+#define HASH_SIZE 32
+#define SIZE_STR_VAR_DB_PKG 12
+
+//private
+
+//data
+typedef struct cur_pkg_tree_node {
+  char *key;
+  char *hash_buffer;
+  char *package_name;
+  unsigned int safe_to_free_package_name;
+  struct cur_pkg_tree_node *greater;
+  struct cur_pkg_tree_node *minor;
+}cur_pkg_tree_node;
+
+
+//functions
+static unsigned int conv_char_int(char dig)
+{
+  if((int) dig > 57) 
+  {
+    return (int)dig - 87;
+  }
+  return (int )dig - 48;
+}
+
+static int compare_hash_num(char *hash1,char*hash2)
+{
+  int temp1,temp2;
+  for(int i=0;i<HASH_SIZE;++i)
+  {
+    temp1=conv_char_int(hash1[i]);
+    temp2=conv_char_int(hash2[i]);
+    if(temp2 > temp1)
+    {
+      return 1;
+    }else if (temp2 < temp1) {
+      return -1;
+    }
+  }
+  return 0;
+}
+
+static char *get_fullname_package(depend_atom *datom)
+{
+  int cat_len,name_len=0;
+  char *package_name =NULL;
+
+  assert(datom!=NULL);
+  assert(datom->CATEGORY!=NULL);
+  assert(datom->PN!=NULL);
+
+  cat_len=strlen(datom->CATEGORY);
+  name_len=strlen(datom->PN);
+  package_name=calloc((cat_len +1+ name_len +1), sizeof(*package_name));
+  package_name[cat_len + 1 +name_len]='\0';
+
+  strcat(package_name,datom->CATEGORY);
+  strcat(package_name,"/");
+  strcat(package_name,datom->PN);
+
+  return package_name;
+}
+static void add_node(cur_pkg_tree_node **root,char *data,char *key,
+                     char *package_name,unsigned int safe_to_free)
+{
+  if(*root==NULL)
+  {
+    *root=xmalloc(sizeof(**root));
+    (*root)->key=key;
+    (*root)->hash_buffer=data;
+    (*root)->package_name=package_name;
+    (*root)->safe_to_free_package_name=safe_to_free;
+    (*root)->greater=NULL;
+    (*root)->minor=NULL;
+    return;
+  }
+
+  int is_greater=compare_hash_num((*root)->key,key);
+  
+  switch (is_greater) {
+    case 0:
+          printf("there are two packages wich update the same file %s %s, the hash of the file is %s\n"
+                 ,package_name,(*root)->package_name,data);
+      return ;
+    case 1:
+      return add_node(&(*root)->greater,data,key,package_name,safe_to_free);
+    case -1:
+      return add_node(&(*root)->minor,data,key,package_name,safe_to_free);
+  }
+}
+
+static char *hash_from_file(char *file_path_complete)
+{
+  FILE *file_to_hash;
+  char buf[512];
+  unsigned char hex_hash[HASH_SIZE+1];
+  hex_hash[HASH_SIZE]='\0';
+  MD5_CTX ctx;
+  
+  char *out = NULL;
+  out=xmalloc(HASH_SIZE+1* sizeof(*out));
+  out[HASH_SIZE]='\0';
+
+  file_to_hash = fopen(file_path_complete,"r");
+  if(file_to_hash == NULL)
+  {
+    fprintf(stderr, "%s not found\n",file_path_complete);
+    out=xmalloc(3* sizeof(*out));
+    out[0]='-';
+    out[1]='1';
+    out[2]='\0';
+
+    fclose(file_to_hash);
+
+    return out;
+  }
+  MD5_Init(&ctx);
```

Comment:
really, there's a libq/hash that provides a several functions for you to reuse instead of re-implementing this

Reply:
Yeah, i see that recently. Recently i was checking to see if i could improve the patch and i notice that. I already changed it in my repo